### ISSUE TITLE
haproxy drain issues

### ISSUE THREAD

**Issue by aemneina** [rancher/rancher#11172]
**Rancher versions:**
rancher/server: v1.6.14
rancher/agent: v1.2.9

**Infrastructure Stack versions:**
healthcheck: v0.3.3-1
ipsec: v0.2.2
network-services: v0.2.8
scheduler: v0.8.3

**Docker version: (`docker version`,`docker info` preferred)**
1.12.6

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
Ubuntu 16.04.1 LTS (4.4.0) 

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
aws

**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
HA

**Environment Template: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle

**Steps to Reproduce:**
1. Deploy a load balancer (with connection draining enabled) fronting two services. 
2. send traffic to both services
3. upgrade one of the two services. (service one in this example)

**Results:**
1. service1 client does gets a:
`HTTP/1.0 503 Service Unavailable`
2. service2 client can see a blip as well:
`curl: (56) Recv failure: Connection reset by peer`

*docker-compose.yml for test stack*
```
version: '2'
services:
  service2:
    image: sangeetha/testhostrouting
    environment:
      asd: dddd
    stdin_open: true
    tty: true
    labels:
      io.rancher.container.pull_image: always
      svc: service2
  service1:
    image: sangeetha/testhostrouting
    environment:
      asd: '123'
    stdin_open: true
    tty: true
    labels:
      io.rancher.container.pull_image: always
      svc: service1
  lbme:
    image: rancher/lb-service-haproxy:v0.7.15
    ports:
    - 80:80/tcp
    labels:
      io.rancher.container.agent.role: environmentAdmin,agent
      io.rancher.container.agent_service.drain_provider: 'true'
      io.rancher.container.create_agent: 'true'
```

*rancher-compose.yml for test stack*
```
version: '2'
services:
  service2:
    scale: 1
    start_on_create: true
    drain_timeout_ms: 10000
  service1:
    scale: 1
    start_on_create: true
    drain_timeout_ms: 10000
  lbme:
    scale: 1
    start_on_create: true
    lb_config:
      certs: []
      config: |-
        defaults
        log 127.0.0.1:8514 local0 debug
      port_rules:
      - hostname: service1.ahmad.rancher.space
        path: ''
        priority: 6
        protocol: http
        selector: svc=service1
        source_port: 80
        target_port: 80
      - hostname: service2.ahmad.rancher.space
        path: ''
        priority: 7
        protocol: http
        selector: svc=service2
        source_port: 80
        target_port: 80
    health_check:
      healthy_threshold: 2
      response_timeout: 2000
      port: 42
      unhealthy_threshold: 3
      initializing_timeout: 60000
      interval: 2000
      reinitializing_timeout: 60000
```

*super awesome test client:*
`while true; do curl -i --header "Host: fqdn" http://<ip_of_host>/name.html ; done`

**Expected Result**
No connection blips on either side.

*edited to enable drain settings on services as well as lb

**Comment by prachidamle** [comment]
@aemneina I have not tried to reproduce this, but just looking at the compose files above, I see that the 'drain_timeout_ms' parameter is missing in rancher_compose.yml for the two services. Setting this parameter to a non-zero time in milliseconds will enable draining of connections. Without this, no drain happens.

http://rancher.com/docs/rancher/v1.6/en/cattle/adding-load-balancers/#draining-target-connections-for-a-load-balancer

**Comment by prachidamle** [comment]
"Draining" of existing connections is best seen using services that hold the connection for some time before responding. 

The image for the services above responds immediately, so there are no long living connections when the service is upgraded. Thus during upgrade, when we check if there are any connection in process to this service, we most of the time do not find any, so no draining is needed and upgrade proceeds.

Our haproxy drain feature supports "draining" or "waiting" for existing connections found while upgrading, if nothing is found we proceed. New connections will be denied while upgrade is in process.

Btw, to not see any service disruption like this during upgrade, we recommend, service scale >1 and upgrade service in batches using "start before stop"

**Comment by prachidamle** [comment]
@deniseschannon discussed the above with @aemneina - we will be checking back with the customer with the above info.

**Comment by prachidamle** [comment]
```service2 client can see a blip as well:
curl: (56) Recv failure: Connection reset by peer```

Reading haproxy 1.8 feature, it seems like the error service2 receives above can be resolved by haproxy 1.8


However it will be still possible to get a 503 for service1 
(when during upgrade we stop the backend container, haproxy's reload wont help here when the backend suddenly goes down)

```for service1, to not receive the 503's we will still need:
1) deploy service with scale>1 and upgrade in batches, possibly with start-before-stop
2) Also have our drain feature for services holding connections long```

**Comment by sangeethah** [comment]
@aemneina , I have not been able to reproduce this issue when testing with latest 1.6 build and haproxy version - rancher/lb-service-haproxy:v0.7.15

In above mentioned scenario, we do see the "503 Service Unavailable" error for the time when service1 is being upgraded when LB is accessing service1 instances which is expected.

I do not see the LB url accessing service2 instances have any issues . They continue to get to "200 OK" response when service1 is being upgraded.


Also tried the same use case with rancher/lb-service-haproxy:v0.9.0 which is the latest LB version. 
 We see the same behavior with this Lb version as well.
